conn_browser.go

```diff
@@ -0,0 +1,229 @@
+		closeSignal:  make(chan struct{}),
+		dataSignal:   make(chan struct{}),
+	}
+	conn.setUpHandlers()
+	go func() {
+		// TODO(albrow): Handle error appropriately
+		err := conn.readLoop()
+		if err != nil {
+			panic(err)
+		}
+	}()
+	return conn
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	if err := c.checkOpen(); err != nil {
+		return 0, io.EOF
+	}
+
+	for {
+		c.currDataMut.RLock()
+		n, err := c.currData.Read(b)
+		c.currDataMut.RUnlock()
+		if err != nil && err != io.EOF {
+			// Return any unexpected errors immediately.
+			return n, err
+		} else if n == 0 || err == io.EOF {
+			// There is no data ready to be read. Wait for more data or for the
+			// connection to be closed.
+			select {
+			case <-c.dataSignal:
+				continue
+			case <-c.closeSignal:
+				return 0, io.EOF
+			}
+		} else {
+			return n, err
+		}
+	}
+}
+
+// checkOpen returns an error if the connection is not open. Otherwise, it
+// returns nil.
+func (c *Conn) checkOpen() error {
+	state := c.Get("readyState").Int()
+	switch state {
+	case webSocketStateClosed, webSocketStateClosing:
+		return errConnectionClosed
+	}
+	return nil
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	if err := c.checkOpen(); err != nil {
+		return 0, err
+	}
+	uint8Array := js.Global().Get("Uint8Array").New(len(b))
+	for i := 0; i < len(b); i++ {
+		uint8Array.SetIndex(i, b[i])
+	}
+	c.Call("send", uint8Array.Get("buffer"))
+	return len(b), nil
+}
+
+// Close closes the connection. Only the first call to Close will receive the
+// close error, subsequent and concurrent calls will return nil.
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.Call("close")
+	if c.messageHandler != nil {
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+	}
+	return nil
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	return NewAddr("0.0.0.0:0")
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	rawURL := c.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) SetDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetReadDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetWriteDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) setUpHandlers() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		// Message handlers already created. Nothing to do.
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			// TODO(albrow): Currently we assume data is of type Blob. Really we
+			// should check binaryType and then decode accordingly.
+			blob := args[0].Get("data")
+			data := readBlob(blob)
+			c.incomingData <- data
```

Comment:
This will still deliver messages out of order as goroutines can be run in any order.

Reply:
The idea behind the read loop was that it would prevent writing to this channel from blocking. That plus a significant buffer would get rid of the need for a goroutine.

However, this could also be accomplished by a lock with a small critical section.

Reply:
I'd go with the lock. It'll be a lot simpler.

Reply:
Sorry @Stebalien I meant for this part to *not* be in a goroutine which would retain the order of messages. I just forgot to remove the goroutine.

Let me try a version where we write to a mutex-protected buffer directly inside of `messageHandler`. I agree it is probably the cleanest solution.